File: app/src/main/kotlin/com/saatxi/eatapp/data/local/Cuisine.kt
package com.saatxi.eatapp.data.local

import androidx.annotation.StringRes
import com.saatxi.eatapp.R

/**
 * The closed vocabulary of cuisine types the app understands.
 *
 * The `restaurants.cuisineType` column of the synced `.db` stores [key] — a
 * stable, language-independent identifier — never a display label. Only
 * [labelRes] is ever translated, so adding a language means adding a
 * `values-xx/strings.xml` and nothing else: the data file is never touched.
 *
 * The vocabulary is documented for data authors in the README's
 * "Data source & updating restaurant data" section. Keep the two in sync.
 */
enum class Cuisine(val key: String, @StringRes val labelRes: Int) {
    // Origin
    MEDITERRANEAN("mediterranean", R.string.cuisine_mediterranean),
    SPANISH("spanish", R.string.cuisine_spanish),
    CATALAN("catalan", R.string.cuisine_catalan),
    BASQUE("basque", R.string.cuisine_basque),
    ITALIAN("italian", R.string.cuisine_italian),
    JAPANESE("japanese", R.string.cuisine_japanese),
    CHINESE("chinese", R.string.cuisine_chinese),
    ASIAN("asian", R.string.cuisine_asian),
    INDIAN("indian", R.string.cuisine_indian),
    MIDDLE_EASTERN("middle_eastern", R.string.cuisine_middle_eastern),
    AMERICAN("american", R.string.cuisine_american),
    SEAFOOD("seafood", R.string.cuisine_seafood),

    // Venue / meal type
    BAR("bar", R.string.cuisine_bar),
    BEER_BAR("beer_bar", R.string.cuisine_beer_bar),
    WINE_BAR("wine_bar", R.string.cuisine_wine_bar),
    CAFE("cafe", R.string.cuisine_cafe),
    BAKERY("bakery", R.string.cuisine_bakery),
    DESSERT("dessert", R.string.cuisine_dessert),
    BREAKFAST("breakfast", R.string.cuisine_breakfast),
    BRUNCH("brunch", R.string.cuisine_brunch),
    GRILL("grill", R.string.cuisine_grill),
    FAST_FOOD("fast_food", R.string.cuisine_fast_food),
    FINE_DINING("fine_dining", R.string.cuisine_fine_dining),
    VEGETARIAN("vegetarian", R.string.cuisine_vegetarian);

    companion object {
        private val byKey = entries.associateBy { it.key }

        /**
         * Resolves a raw `cuisineType` value, or null when the file uses a key
         * this build doesn't know. Callers must degrade gracefully rather than
         * fail: an unrecognised key still shows the raw string and a generic
         * icon, so a newer data file never breaks an older app.
         */
        fun fromKey(key: String?): Cuisine? =
            key?.trim()?.lowercase()?.let(byKey::get)
    }
}

File: lib/data/models/cuisine.dart
import '../../core/l10n/generated/app_localizations.dart';

/// The closed vocabulary of cuisine types the app understands.
///
/// The `restaurants.cuisineType` column stores [key] — a stable,
/// language-independent identifier — never a display label. Only [label] is
/// ever translated, so adding a language means adding an ARB file and nothing
/// else.
///
/// The vocabulary is documented in the README. Keep the two in sync.
enum Cuisine {
  // Origin
  mediterranean('mediterranean'),
  spanish('spanish'),
  catalan('catalan'),
  basque('basque'),
  italian('italian'),
  japanese('japanese'),
  chinese('chinese'),
  asian('asian'),
  indian('indian'),
  middleEastern('middle_eastern'),
  american('american'),
  seafood('seafood'),

  // Venue / meal type
  bar('bar'),
  beerBar('beer_bar'),
  wineBar('wine_bar'),
  cafe('cafe'),
  bakery('bakery'),
  dessert('dessert'),
  breakfast('breakfast'),
  brunch('brunch'),
  grill('grill'),
  fastFood('fast_food'),
  fineDining('fine_dining'),
  vegetarian('vegetarian');

  const Cuisine(this.key);

  /// Stable, language-independent identifier stored in the database.
  final String key;

  /// The translated display label for this cuisine.
  String label(AppLocalizations l10n) => switch (this) {
    Cuisine.mediterranean => l10n.cuisineMediterranean,
    Cuisine.spanish => l10n.cuisineSpanish,
    Cuisine.catalan => l10n.cuisineCatalan,
    Cuisine.basque => l10n.cuisineBasque,
    Cuisine.italian => l10n.cuisineItalian,
    Cuisine.japanese => l10n.cuisineJapanese,
    Cuisine.chinese => l10n.cuisineChinese,
    Cuisine.asian => l10n.cuisineAsian,
    Cuisine.indian => l10n.cuisineIndian,
    Cuisine.middleEastern => l10n.cuisineMiddleEastern,
    Cuisine.american => l10n.cuisineAmerican,
    Cuisine.seafood => l10n.cuisineSeafood,
    Cuisine.bar => l10n.cuisineBar,
    Cuisine.beerBar => l10n.cuisineBeerBar,
    Cuisine.wineBar => l10n.cuisineWineBar,
    Cuisine.cafe => l10n.cuisineCafe,
    Cuisine.bakery => l10n.cuisineBakery,
    Cuisine.dessert => l10n.cuisineDessert,
    Cuisine.breakfast => l10n.cuisineBreakfast,
    Cuisine.brunch => l10n.cuisineBrunch,
    Cuisine.grill => l10n.cuisineGrill,
    Cuisine.fastFood => l10n.cuisineFastFood,
    Cuisine.fineDining => l10n.cuisineFineDining,
    Cuisine.vegetarian => l10n.cuisineVegetarian,
  };

  /// Resolves a raw `cuisineType` value, or null when the data uses a key this
  /// build doesn't know. Callers must degrade gracefully rather than fail: an
  /// unrecognised key still shows the raw string and a generic icon, so a
  /// newer data file never breaks an older app.
  static Cuisine? fromKey(String? key) => switch (key?.trim().toLowerCase()) {
    'mediterranean' => Cuisine.mediterranean,
    'spanish' => Cuisine.spanish,
    'catalan' => Cuisine.catalan,
    'basque' => Cuisine.basque,
    'italian' => Cuisine.italian,
    'japanese' => Cuisine.japanese,
    'chinese' => Cuisine.chinese,
    'asian' => Cuisine.asian,
    'indian' => Cuisine.indian,
    'middle_eastern' => Cuisine.middleEastern,
    'american' => Cuisine.american,
    'seafood' => Cuisine.seafood,
    'bar' => Cuisine.bar,
    'beer_bar' => Cuisine.beerBar,
    'wine_bar' => Cuisine.wineBar,
    'cafe' => Cuisine.cafe,
    'bakery' => Cuisine.bakery,
    'dessert' => Cuisine.dessert,
    'breakfast' => Cuisine.breakfast,
    'brunch' => Cuisine.brunch,
    'grill' => Cuisine.grill,
    'fast_food' => Cuisine.fastFood,
    'fine_dining' => Cuisine.fineDining,
    'vegetarian' => Cuisine.vegetarian,
    _ => null,
  };
}
